zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1008 @@
+
+        if possible_scalar.is_some().into() {
+            Ok(Self(possible_scalar.unwrap()))
+        } else {
+            Err("Invalid pallas::Scalar value")
+        }
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard incoming
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardinviewingkeyencoding
+mod ivk_hrp {
+    pub const MAINNET: &str = "zivko";
+    pub const TESTNET: &str = "zivktestorchard";
+}
+
+/// An incoming viewing key, as described in [protocol specification
+/// §4.2.3][ps].
+///
+/// Used to decrypt incoming notes without spending them.
+///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct IncomingViewingKey {
+    network: Network,
+    scalar: pallas::Scalar,
+}
+
+// TODO: impl a From that accepts a Network?
+
+impl fmt::Debug for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("IncomingViewingKey")
+            .field(&hex::encode(self.scalar.to_bytes()))
+            .finish()
+    }
+}
+
+impl fmt::Display for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let hrp = match self.network {
+            Network::Mainnet => ivk_hrp::MAINNET,
+            _ => ivk_hrp::TESTNET,
+        };
+
+        bech32::encode_to_fmt(f, hrp, &self.scalar.to_bytes().to_base32(), Variant::Bech32).unwrap()
+    }
+}
+
+impl From<[u8; 32]> for IncomingViewingKey {
+    /// Generate an _IncomingViewingKey_ from existing bytes.
+    fn from(bytes: [u8; 32]) -> Self {
+        Self {
+            // TODO: handle setting the Network better.
+            network: Network::default(),
+            scalar: pallas::Scalar::from_bytes(&bytes).unwrap(),
+        }
+    }
+}
+
+impl From<FullViewingKey> for IncomingViewingKey {
+    /// Commit^ivk_rivk(ak, nk) :=
+    ///    SinsemillaShortCommit_rcm (︁"z.cash:Orchard-CommitIvk", I2LEBSP_l(ak) || I2LEBSP_l(nk)︁) mod r_P
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteprfs
+    #[allow(non_snake_case)]
+    fn from(fvk: FullViewingKey) -> Self {
+        let mut M: BitVec<Lsb0, u8> = BitVec::new();
+
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.spend_validating_key)[..],
+        ));
+        M.append(&mut BitVec::<Lsb0, u8>::from_slice(
+            &<[u8; 32]>::from(fvk.nullifier_deriving_key)[..],
+        ));
+
+        // Commit^ivk_rivk
+        let commit_x = sinsemilla_short_commit(
+            fvk.ivk_commit_randomness.into(),
+            b"z.cash:Orchard-CommitIvk",
+            &M,
+        );
+
+        Self {
+            // TODO: handle the network better, maybe an enum variant constraint?
+            network: Network::default(),
```

Comment:
Rename to `IncomingViewingKey::from_full` and add a `Network` argument.

Reply:
Resolved in dda853829343ee863477c1f74aefda4f402044ae